src/ray/object_manager/object_manager.h

```diff
@@ -135,132 +148,129 @@ class ObjectManager {
   ray::Status Terminate();
 
  private:
-  using BoostEC = const boost::system::error_code &;
-
   ClientID client_id_;
   ObjectManagerConfig config_;
   std::unique_ptr<ObjectDirectoryInterface> object_directory_;
-  std::unique_ptr<ObjectStoreClient> store_client_;
+  std::unique_ptr<ObjectStoreNotification> store_notification_;
+  std::unique_ptr<ObjectStorePool> store_pool_;
 
   /// An io service for creating connections to other object managers.
-  boost::asio::io_service io_service_;
+  std::unique_ptr<boost::asio::io_service> object_manager_service_;
+  /// Weak reference to main service. We ensure this object is destroyed before
+  /// main_service_ is stopped.
+  boost::asio::io_service *main_service_;
```

Comment:
Why a reference instead of a pointer? This should be fine as an observational pointer (see https://herbsutter.com/elements-of-modern-c-style/). The owner of `main_service_` outlives the object manager, and the object manager has no control over the lifetime of `main_service_`.